pkg/kv/kvserver/store_raft.go

```diff
@@ -42,47 +55,177 @@ type raftRequestInfo struct {
+		// actually queued, but the reverse would be a serious bug leading to
+		// indefinitely queued messages.
+		queuedAtRaftScheduler bool
+
+		// State managed by the storeRaftDequeuePacer.
+		//
+		// The raftReceiveQueue and storeRaftDequeuePacer interact via three
+		// fields in the raftReceiveQueue. One of these is pausedMsgStreamAcc
+		// which is read by storeRaftDequeuePacer and modified by raftReceiveQueue
+		// (also mentioned above). The other two are the following pacerDequeue*
+		// fields that are read and modified by the pacer. The
+		// pacerDequeueDecision is read by raftReceiveQueue to decide on its
+		// actions. The pacerDequeueAccountingBytes is used for bookkeeping by the
+		// pacer, and set to 0 when the MsgApps are drained from the
+		// pausedMsgStream (it is distinct from pausedMsgStreamAcc.Used() as more
+		// MsgApps can be queued to that stream after the pacer switches the
+		// decision to raftDequeueAll).
+		//
+		// INVARIANTS:
+		//
+		// pacerDequeueDecision == raftDequeueSkipPaused <=>
+		//   pacerDequeueAccountingBytes == 0
+		//
+		// pacerDequeueDecision == raftDequeueSkipPaused <=>
+		//   the raftReceiveQueue is enqueued in the pacer &&
+		//   len(pausedMsgStream) > 0
+		//
+		// The first invariant is trivially maintained by ensuring that any
+		// transition from raftDequeueAll => raftDequeueSkipPaused only happens
+		// when the pausedMsgStream is empty and a MsgApp is received.
+		//
+		// The transition from raftDequeueSkipPaused => raftDequeueAll happens
+		// when the store is healthy enough (decided by the storeRaftDequeuePacer)
+		// for the pausedMsgStream to be drained.
+		//
+		// See the declaration of storeRaftDequeuePacer for the methods called on
+		// the pacer by the raftReceiveQueue.
+		pacerDequeueDecision        raftDequeueDecision
+		pacerDequeueAccountingBytes int64
 	}
 	maxLen int
-	acc    mon.BoundAccount
+}
+
+func newRaftReceiveQueue(
+	rangeID roachpb.RangeID, pacer *storeRaftDequeuePacer, monitor *mon.BytesMonitor, maxLen int,
+) *raftReceiveQueue {
+	q := &raftReceiveQueue{rangeID: rangeID, dequeuePacer: pacer, maxLen: maxLen}
+	q.mu.pausedMsgStreamAcc.Init(context.Background(), monitor)
+	q.mu.normalMsgStreamAcc.Init(context.Background(), monitor)
+	return q
 }
 
 // Len returns the number of requests in the queue.
 func (q *raftReceiveQueue) Len() int {
 	q.mu.Lock()
 	defer q.mu.Unlock()
-	return len(q.mu.infos)
+	return q.lenLocked()
+}
+
+func (q *raftReceiveQueue) lenLocked() int {
+	return len(q.mu.normalMsgStream) + len(q.mu.pausedMsgStream)
 }
 
-// Drain moves the stored requests out of the queue, returning them to
-// the caller. Returns true if the returned slice was not empty.
+// Drain moves the stored requests out of the queue, returning them to the
+// caller. Returns true if the returned slice was not empty. It will always
+// drain the normalMsgStream, and will also drain the pausedMsgStream if the
+// (internal) raftDequeueDecision permits.
```

Comment:
Pre-existing: is there sense in this `bool` return parameter? Looks like `bool` is true iff `len([]raftRequestInfo) != 0`?

Would be nice to fix by a small commit preceding the main stuff.